### PR TITLE
Basic/initial Mixcloud support

### LINKED ISSUE #568: Mixcloud Support
**Is your feature request related to a problem? Please describe.**
N/A

**Describe the solution you'd like**
Last one, I swear. 😁  After the successful implementation of #544, there is one last provider that I would like to see implemented, and that is Mixcloud:
https://www.mixcloud.com/developers/#authorization

**Describe alternatives you've considered**
You already have SoundCloud (which is fantastic!), and I have some potential customers that use/prefer Mixcloud as well.  It would be great to allow them access via your terrific integration work here.

**Additional context**
Thank you again for all your impressive and diligent work over there. 🙏

### PR DESCRIPTION
**Pull request by Mike-E-angelo**
This is in reference to #568.

I was able to get things mostly working, and it does work with a live connection and tests.  However, it appears that Mixcloud does not support the `state` parameter and does not pass this back to the handler.  So what I have done is bundle the `state` in the `redirect_uri` and then unbundle it when it returns.  Sort of ugly, but works.

I am definitely open to feedback and suggestions on improving this.  Let the review begin. 😁

### CODE REVIEW

File: src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+using System;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace AspNet.Security.OAuth.Mixcloud
+{
+    /// <summary>
+    /// Extension methods to add  authentication capabilities to an HTTP application pipeline.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationExtensions.cs]:
```suggestion
    /// Extension methods to add Mixcloud authentication capabilities to an HTTP application pipeline.
```

File: src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        public MixcloudAuthenticationHandler(
+            [NotNull] IOptionsMonitor<MixcloudAuthenticationOptions> options,
+            [NotNull] ILoggerFactory logger,
+            [NotNull] UrlEncoder encoder,
+            [NotNull] ISystemClock clock)
+            : base(options, logger, encoder, clock)
+        {
+        }
+
+        protected override string BuildChallengeUrl(AuthenticationProperties properties, string redirectUri)
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationHandler.cs]:
```suggestion
        protected override string BuildChallengeUrl([NotNull] AuthenticationProperties properties, [NotNull] string redirectUri)
```

File: src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        {
+            var scopeParameter = properties.GetParameter<ICollection<string>>(OAuthChallengeProperties.ScopeKey);
+            var scope = scopeParameter != null ? FormatScope(scopeParameter) : FormatScope();
+
+            var parameters = new Dictionary<string, string?>
+            {
+                { "client_id", Options.ClientId },
+                { "scope", scope },
+                { "response_type", "code" }
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationHandler.cs]:
```suggestion
                ["client_id"] = Options.ClientId,
                ["scope"] = scope,
                ["response_type"] = "code"
```

**Comment by Mike-E-angelo** [comment]
Mac not having it. 😆😭  It can sense my internal hatred of it. :P

I am not entirely sure of the issue with the build breaking but can assist if needed.

File: src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+            }
+
+            var state = Options.StateDataFormat.Protect(properties);
+            parameters["state"] = state;
+            parameters["redirect_uri"] = QueryHelpers.AddQueryString(redirectUri, "state", state);
+
+            return QueryHelpers.AddQueryString(Options.AuthorizationEndpoint, parameters);
+        }
+
+        protected override async Task<HandleRequestResult> HandleRemoteAuthenticateAsync()
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationHandler.cs:77]: All this code rings a bell, but what's the main reason with the Mixcloud provider the override is needed (compared to the default implementation in the base class)?

File: src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationOptions.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+{
+    /// <summary>
+    /// Defines a set of options used by <see cref="MixcloudAuthenticationHandler"/>.
+    /// </summary>
+    public class MixcloudAuthenticationOptions : OAuthOptions
+    {
+        public MixcloudAuthenticationOptions()
+        {
+            ClaimsIssuer = MixcloudAuthenticationDefaults.Issuer;
+
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationOptions.cs]:
```suggestion
```

File: test/AspNet.Security.OAuth.Providers.Tests/Mixcloud/MixcloudTests.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        public override string DefaultScheme => MixcloudAuthenticationDefaults.AuthenticationScheme;
+
+        protected internal override void RegisterAuthentication(AuthenticationBuilder builder)
+        {
+            builder.AddMixcloud(options => ConfigureDefaults(builder, options));
+        }
+
+        [Theory]
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/Mixcloud/MixcloudTests.cs:29]: Missing a test case for the bio claim.

File: src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationConstants.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    /// Contains constants specific to the <see cref="MixcloudAuthenticationHandler"/>.
+    /// </summary>
+    public static class MixcloudAuthenticationConstants
+    {
+        public static class Claims
+        {
+            public const string FullName = "urn:mixcloud:fullname";
+            public const string ProfileUrl = "urn:mixcloud:profileurl";
+            public const string City = "urn:mixcloud:city";
+            public const string Bio = "urn:mixcloud:bio";
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationConstants.cs]: Would `Biography` be a better name for the constant?

**Review by martincostello — COMMENTED** [review]
Also please update the table in the README. There's a comment at the end of the table you can copy to add a new row.

**Comment by martincostello** [comment]
> I am not entirely sure of the issue with the build breaking but can assist if needed.

#570

File: src/AspNet.Security.OAuth.Mixcloud/AspNet.Security.OAuth.Mixcloud.csproj [hunk 8]
```diff
@@ -0,0 +1,18 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>net5.0</TargetFrameworks>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <Description>ASP.NET Core security middleware enabling Mixcloud authentication.</Description>
+    <Authors>Adam Reisinger;Chino Chang;Michael DeMond</Authors>
```
Comment (kinosang) [inline src/AspNet.Security.OAuth.Mixcloud/AspNet.Security.OAuth.Mixcloud.csproj]: No need to sign my name here, bro.

File: src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationHandler.cs (hunk 4 above)
Comment (Mike-E-angelo) [reply src/AspNet.Security.OAuth.Mixcloud/MixcloudAuthenticationHandler.cs:77]: This is for the bundling/unbundling of `state` as Mixcloud does not appear to support this.  I have added comments in the code to reflect this intention.

File: src/AspNet.Security.OAuth.Mixcloud/AspNet.Security.OAuth.Mixcloud.csproj (hunk 8 above)
Comment (Mike-E-angelo) [reply src/AspNet.Security.OAuth.Mixcloud/AspNet.Security.OAuth.Mixcloud.csproj]: Haha good point...  I copied/pasted your code straight up, so felt the need to attribute accordingly.  I will remove.

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
This looks good to me. If the tests on macOS keep failing you might need to rebase once I fix that test and merge it to `dev`.

Also, have you validated that this provided works correctly against the real Mixcloud service? We've had a few PRs recently where I've assumed people have tested new providers with the relevant services, but haven't, and then they have had to submit PRs to fix them after they've been shipped to NuGet.org because they didn't work.

**Comment by Mike-E-angelo** [comment]
> This looks good to me. If the tests on macOS keep failing you might need to rebase once I fix that test and merge it to dev.

Great!  Thank you for your quick collaboration.  And no problem about rebasing.  Looks like we have more green already. 🤞

> Also, have you validated that this provided works correctly against the real Mixcloud service?

That's actually how I started out. :)  I got it working live and then got those tests passing.  Then I tested live again just to make sure. :P

FWIW, I can be on-call if someone reports a problem.  I'm building my application on this sucker so I'm a bit motivated to make sure it works well.  If needed, send out a mention and I will look into it for you. 👍

BTW/FWIWx2: you can find a little bit of what I am working on here, with a shout out to this project a little ways down for the Coinbase support:
https://blog.starbeam.one/2021/06/june-2021-status-update/

**Comment by martincostello** [comment]
Great, thanks @Mike-E-angelo - I'll merge this now and look at shipping it as part of the 5.0.7 release at some point this weekend.

Until then, you'll be able to consume the provider as a pre-release version from our MyGet feed.

**Pull request merged**